lib/grandpa/grandpa.go

```diff
@@ -38,178 +50,192 @@ func NewService(blockState BlockState, voters []*Voter) (*Service, error) {
+// if there are no blocks that have >=2/3 direct votes, this function will find predecessors of those blocks that do have >=2/3 votes.
+// note that by voting for a block, all of its predecessor blocks are automatically voted for.
+// thus, if there are no blocks with >=2/3 total votes, but the sum of votes for blocks A and B is >=2/3, then this function returns
+// the first common predecessor of A and B.
+func (s *Service) getPossiblePreVotedBlocks() (map[common.Hash]uint64, error) {
+	// get blocks that were directly voted for
+	votes := s.getDirectVotes()
+	blocks := make(map[common.Hash]uint64)
+
+	// check if any of them have >=2/3 votes
+	for v := range votes {
+		total, err := s.getTotalVotesForBlock(v.hash)
+		if err != nil {
+			return nil, err
+		}
+
+		if total >= uint64(2*len(s.state.voters)/3) {
+			blocks[v.hash] = v.number
+		}
 	}
 
-	// check for equivocation ie. multiple votes within one subround
-	voter, err := s.state.pubkeyToVoter(pk)
-	if err != nil {
-		return nil, err
+	// since we want to select the block with the highest number that has >=2/3 votes,
+	// we can return here since their predecessors won't have a higher number.
+	if len(blocks) != 0 {
+		return blocks, nil
 	}
 
-	vote := NewVote(m.message.hash, m.message.number)
+	// no block has >=2/3 direct votes, check for votes for predecessors recursively
+	var err error
+	va := s.getVotes()
 
-	equivocated := s.checkForEquivocation(voter, vote)
-	if equivocated {
-		return nil, ErrEquivocation
+	for v := range votes {
+		blocks, err = s.getPossiblePreVotedPredecessors(va, v.hash, blocks)
+		if err != nil {
+			return nil, err
+		}
 	}
 
-	err = s.validateVote(vote)
-	if err != nil {
-		return nil, err
-	}
+	return blocks, nil
+}
 
-	s.votes[pk.AsBytes()] = vote
+// getPossiblePreVotedPredecessors recursively searches for predecessors with >=2/3 votes
+// it returns a map of block hash -> number, such that the blocks in the map have >=2/3 votes
+func (s *Service) getPossiblePreVotedPredecessors(votes []Vote, curr common.Hash, prevoted map[common.Hash]uint64) (map[common.Hash]uint64, error) {
+	for _, v := range votes {
+		if v.hash == curr {
+			continue
+		}
 
-	return vote, nil
-}
+		// find common predecessor, check if votes for it is >=2/3 or not
+		pred, err := s.blockState.HighestCommonAncestor(v.hash, curr)
+		if err != nil {
+			return nil, err
+		}
 
-func validateMessageSignature(pk *ed25519.PublicKey, m *VoteMessage) error {
-	msg, err := scale.Encode(&FullVote{
-		stage: m.stage,
-		vote:  NewVote(m.message.hash, m.message.number),
-		round: m.round,
-		setID: m.setID,
-	})
-	if err != nil {
-		return err
-	}
-	ok, err := pk.Verify(msg, m.message.signature[:])
-	if err != nil {
-		return err
-	}
+		if pred == curr {
+			return prevoted, nil
+		}
 
-	if !ok {
-		return ErrInvalidSignature
+		total, err := s.getTotalVotesForBlock(pred)
+		if err != nil {
+			return nil, err
+		}
+
+		if total >= uint64(2*len(s.state.voters)/3) {
```

Comment:
uint64(2*len(s.state.voters)/3) could be a var and named with something that gives a indication of why this is here for. The current if condition is somewhat vague without it.

Reply:
the function comment describes it, but I can move it to a var